src/iconv.jl

```diff
@@ -1,9 +1,54 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
 module iconv
-import Base: close, eof, flush, read, readall, write
+import Base: close, eof, flush, read, readall, write, show
 import Base.Libc: errno, strerror, E2BIG, EINVAL, EILSEQ
 export StringEncoder, StringDecoder, encode, decode
+export StringEncodingError, OutputBufferError, IConvError
+export InvalidEncodingError, InvalidSequenceError, IncompleteSequenceError
+
+
+abstract StringEncodingError
+
+# specified encodings are not supported
+type InvalidEncodingError <: StringEncodingError
+    args::Tuple{ASCIIString, ASCIIString}
+    InvalidEncodingError(from, to) = new((ascii(from), ascii(to)))
+end
+message(::Type{InvalidEncodingError}) = "Conversion from <<1>> to <<2>> not supported by iconv implementation, check that specified encodings are correct"
+
+# Encountered invalid byte sequence
+type InvalidSequenceError <: StringEncodingError
+    args::Tuple{ASCIIString}
+    InvalidSequenceError(seq) = new((bytes2hex(seq),))
+end
+message(::Type{InvalidSequenceError}) = "Byte sequence 0x<<1>> is invalid in source encoding or cannot be represented in target encoding"
+
+type IConvError <: StringEncodingError
+    args::Tuple{ASCIIString, Int, ASCIIString}
+    IConvError(func) = new((func, errno(), strerror(errno())))
+end
+message(::Type{IConvError}) = "<<1>>: <<2>> (<<3>>)"
+
+# Input ended with incomplete byte sequence
+type IncompleteSequenceError <: StringEncodingError ; end
+message(::Type{IncompleteSequenceError}) = "Incomplete byte sequence at end of input"
+
+type OutputBufferError <: StringEncodingError ; end
+message(::Type{OutputBufferError}) = "Ran out of space in the output buffer"
+
+type InvalidBytesError <: StringEncodingError ; end
+message(::Type{InvalidBytesError}) = "Invalid byte sequence in input"
+
+function show(io::IO, exc::StringEncodingError)
+    str = message(typeof(exc))
+    for i = 1:length(exc.args)
+        str = replace(str, "<<$i>>", exc.args[i])
+    end
+    print(io, str)
+end
+
+show{T<:Union{IncompleteSequenceError,OutputBufferError,InvalidBytesError}}(io::IO, exc::T) = print(io, message(T))
```

Comment:
Again, no need to use the longer parametric function syntax here, just put the `Union` after the argument (also, putting the function on two lines would be better).